### PR TITLE
feat: Handle QTextEdit in the ScriptDialogItem

### PR DESCRIPTION
**Pull request by smnppKDAB**
Fixes KDAB#27
to use the syntax highlighting of the "data.foo" widget, just use the name of the extension like "cpp", "js" or "py" in the "data.fooSyntax" property

### CODE REVIEW

File: src/core/scriptdialogitem.cpp [hunk 1]
```diff
@@ -552,12 +560,64 @@ void ScriptDialogItem::createProperties(QWidget *dialogWidget)
...
     }
 
     m_data->ready();
 }
 
+void ScriptDialogItem::applySyntaxHighlighting(QTextEdit *textEdit)
+{
+    QString syntax = "." + textEdit->property("textEditSyntax").toString();
+
+    if (!syntax.isEmpty()) {
```
Comment (dfaure-kdab) [inline src/core/scriptdialogitem.cpp:589]: How could it be empty, given that you prepended a "." ?
You have to test for emptiness before prepending "."

File: src/core/scriptdialogitem.cpp [hunk 2]
```diff
@@ -552,12 +560,64 @@ void ScriptDialogItem::createProperties(QWidget *dialogWidget)
...
+        auto highlighter = new KSyntaxHighlighting::SyntaxHighlighter(textEdit->document());
+        static KSyntaxHighlighting::Repository repository;
+        highlighter->setTheme(repository.themeForPalette(QApplication::palette()));
+        const auto definition = repository.definitionForFileName(syntax);
+        if (!definition.isValid()) {
+            qWarning() << "No valid syntax definition";
```
Comment (dfaure-kdab) [inline src/core/scriptdialogitem.cpp:595]: Give more details to facilitate debugging, like "No valid syntax definition for file extension" << syntax;

File: src/core/scriptdialogitem.cpp [hunk 3]
```diff
@@ -593,6 +663,18 @@ void ScriptDialogItem::changeValue(const QString &key, const QVariant &value)
         return;
     }
 
+    // It may be a syntax rule
+    if (key.endsWith("Syntax")) {
+        if (auto textEdit = findChild<QTextEdit *>(key.left(key.length() - 6))) {
+            textEdit->setProperty("textEditSyntax", value.toString());
```
Comment (dfaure-kdab) [inline src/core/scriptdialogitem.cpp:669]: What is this property useful for? Why not just pass the value to the applySyntaxHighlighting function directly?

**Review by dfaure-kdab — CHANGES_REQUESTED** [review]

**Review by dfaure-kdab — CHANGES_REQUESTED** [review]

File: src/core/scriptdialogitem.cpp [hunk 4]
```diff
@@ -552,6 +560,22 @@ void ScriptDialogItem::createProperties(QWidget *dialogWidget)
                 completer->setModel(comboBox->model());
                 comboBox->setCompleter(completer);
             }
+        } else if (auto textEdit = qobject_cast<QTextEdit *>(widget)) {
+            m_data->addProperty(widget->objectName().toLocal8Bit(), "QString", QMetaType::QString,
+                                textEdit->toPlainText());
+            connect(textEdit, &QTextEdit::textChanged, this, [this, textEdit]() {
+                m_data->setProperty(sender()->objectName().toLocal8Bit(), textEdit->toPlainText());
+            });
+            m_data->addProperty((widget->objectName() + "Syntax").toLocal8Bit(), "QString", QMetaType::QString,
```
[the remark this answers is not in the document]
Comment (smnppKDAB) [reply src/core/scriptdialogitem.cpp:569]: it didn't work if i didn't add this line

File: src/core/scriptdialogitem.h [hunk 5]
```diff
@@ -85,6 +90,9 @@ public slots:
     static qsizetype countChildren(QQmlListProperty<QObject> *list);
     static void clearChildren(QQmlListProperty<QObject> *list);
 
+    void applySyntaxHighlighting(QTextDocument *document, const QString &syntax);
+    // void applySyntaxHighlighting(KSyntaxHighlighting::SyntaxHighlighter *highlighter, const QString &syntax);
```
Comment (dfaure-kdab) [inline src/core/scriptdialogitem.h:94]: Remove

File: src/core/scriptdialogitem.cpp [hunk 6]
```diff
@@ -593,6 +627,22 @@ void ScriptDialogItem::changeValue(const QString &key, const QVariant &value)
         return;
     }
 
+    // It may be a syntax rule
+    if (key.endsWith("Syntax")) {
+        if (auto textEdit = findChild<QTextEdit *>(key.left(key.length() - 6))) {
+            QString syntax = value.toString();
+            if (!syntax.isEmpty()) {
```
Comment (dfaure-kdab) [inline src/core/scriptdialogitem.cpp:634]: As I said, you could do that test inside the shared function, now.

File: src/core/scriptdialogitem.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (dfaure-kdab) [reply src/core/scriptdialogitem.cpp:569]: OK, makes sense in fact, thanks.

**Review by dfaure-kdab — CHANGES_REQUESTED** [review]

**Pull request merged**